packages/markerikson-stack-client-prototype/src/features/sources/sourcesCache.ts

```diff
@@ -0,0 +1,118 @@
+  with a `.then` method, plus `.resolve/reject` methods, and is doing
+  manual "pending/resolved" tracking.  This is apparently to avoid the
+  overhead of promises and microtasks.
+
+  RTKQ recently added a `getRunningOperationPromise` method, which returns
+  a consistent promise reference _if_ there's a request in flight for that
+  particular endpoint+args combination.  So, we can use that as the thrown promise.
+
+  I've copy-pasted some of the boolean flag derivation logic from the guts of
+  RTKQ to help figure out if the RTKQ cache entry has data available or not.
+ */
+
+let store: AppStore;
+
+export const setStore = (_store: AppStore) => {
+  store = _store;
+};
+
+// Placeholder for the result given to React's Suspense Cache
+interface SourceGroupsResult {
+  groups: SourceGroups | null;
+}
+
+// We know there's only one cache entry, since it's a "list of things" query.
+// Create a selector to read that cache entry from the Redux store.
+const selectSourceGroupsEntry = api.endpoints.getSources.select();
+
+function createSourceGroupsRecord(): SourceGroupsResult {
+  return {
+    groups: null,
+  };
+}
+
+export function getSourceGroups(): SourceGroups {
+  // Create or read the cache for this section of the UI (????)
+  const sourceGroupsRecord = getCacheForType(createSourceGroupsRecord);
+
+  // Do we have this data already cached? If so, return it
+  if (sourceGroupsRecord.groups !== null) {
+    console.log("Found existing cached sources data, returning");
+    return sourceGroupsRecord.groups;
+  }
+
+  // RTKQ side: have we ever tried to request this data?
+  let sourcesEntry: ReturnType<typeof selectSourceGroupsEntry> | undefined =
+    selectSourceGroupsEntry(store.getState());
+
+  if (!sourcesEntry || sourcesEntry.isUninitialized) {
+    console.log("No RTKQ entry found, initiating request");
+    // If not, actually start fetching
+    store.dispatch(api.endpoints.getSources.initiate());
+
+    // Now we've got an entry in the store
+    sourcesEntry = selectSourceGroupsEntry(store.getState());
+  }
+
+  //Copy-pasted from the guts of RTKQ's `buildHooks.ts`,
+  // since this is derived state
+
+  // data is the last known good request result we have tracked - or if none has been tracked yet the last good result for the current args
+  let data = sourcesEntry!.data;
+
+  const hasData = data !== undefined;
+
+  // isFetching = true any time a request is in flight
+  const isFetching = sourcesEntry!.isLoading;
+  // isLoading = true only when loading while no data is present yet (initial load with no data in the cache)
+  const isLoading = !hasData && isFetching;
+  // isSuccess = true when data is present
+  const isSuccess = sourcesEntry!.isSuccess || (isFetching && hasData);
+
+  if (isFetching || sourcesEntry!.isUninitialized) {
+    console.log("Getting API promise and throwing");
+    const promise = api.util.getRunningOperationPromise("getSources", undefined)!;
+    promise.then(result => {
```

Comment:
I think if the component were to re-request this data before the promise finishes, you'd end up adding multiple `.then()` callbacks, wouldn't you?

Reply:
Hmm. _Can_ that happen? Because the component is suspended and we're waiting for this exact promise to resolve before it will render again.

Reply:
Yes. React might try re-rendering in some scenarios. This is why it's important that the _thing that's thrown_ is stable, for example. (That would also be true of any event handlers you added to that thing.)

Reply:
Gotcha.  On the other hand, in this particular scenario, is that actually a problem?

The `.then` is solely for the purpose of mutating the record with the received value, so if it runs more than once it's a no-op.  Also, we're throwing the original promise, not the `.then` promise.

(I was also having a bit of trouble trying to translate the sync "wakeable" logic in your example over to an async promise, which is how I ended up with this setup.)

Reply:
That's fair.

Unless I'm misunderstanding something, I think this cache could be simplified to the following:
```ts
import { unstable_getCacheForType as getCacheForType } from "react";

import { api, SourceGroups } from "../../app/api";
import type { AppStore } from "../../app/store";

let hasPromiseListenerBeenAttached = false;
let sourceGroups: SourceGroups | null;

const selectSourceGroupsEntry = api.endpoints.getSources.select();

export function getSourceGroups(store: AppStore): SourceGroups {
  if (sourceGroups !== null) {
    return sourceGroups;
  }

  let sourcesEntry: ReturnType<typeof selectSourceGroupsEntry> | undefined =
    selectSourceGroupsEntry(store.getState());

  if (!sourcesEntry || sourcesEntry.isUninitialized) {
    store.dispatch(api.endpoints.getSources.initiate());
    sourcesEntry = selectSourceGroupsEntry(store.getState());
  }

  const promise = api.util.getRunningOperationPromise("getSources", undefined)!;

  if (!hasPromiseListenerBeenAttached) {
    hasPromiseListenerBeenAttached = true;
    promise.then(result => {
      sourceGroups = result.data as SourceGroups;
    });
  }
  
  throw promise;
}
```